### PR TITLE
ref(ui): Converted the Link component from to typescript

### PR DESCRIPTION
**Pull request by priscilawebdev**
**Type:** Refactor

**Description:** 

converted` links.jsx` to `link.tsx`
updated `externalLink.tsx` and `linkWithConfirmation.tsx`

more changes are need but let's do it step by step

### CODE REVIEW

File: src/sentry/static/sentry/app/components/alertLink.tsx [hunk 1]
```diff
@@ -33,10 +33,10 @@ export default class AlertLink extends React.Component<Props> {
     return (
       <StyledLink
         to={(this.props as PropsWithTo).to}
-        href={(this.props as PropsWithHref).href}
         onClick={onClick}
         size={size}
         priority={priority}
+        external
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/alertLink.tsx]: Are alert links always external?

Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/alertLink.tsx]: Well, they were using the `ExternalLink` component before

Comment (markstory) [reply src/sentry/static/sentry/app/components/alertLink.tsx]: Good point. Didn't think of that.

**Review by billyvg — CHANGES_REQUESTED** [review]
I have a few concerns here:

1) We are changing the behavior of `<Link>` - the original intention of `<Link>` was to support navigation through `react-router`/pushState, but also to allow fallback to regular browser navigation for our server-rendered views (e.g. when no router context is found). Now most of our server-rendered views are gone, but there are still some remaining and this can break navigation on those views. We should try evaluate if we still need this behavior before removing it.

2) I'm also worried that `<Link to={}>` and `<Link to={} external>` are too similar. It's easy to confuse or forget the `external` prop and have a broken link. Although we can have the same bugs with the current approach, I think having the distinction via 2 separate components (and prop), helps show that they are two different link types.

**Comment by priscilawebdev** [comment]
@billyvg I understand your concerns, but I also think that the way it is now is more confusing. We are using the Link, ExternalLink, TextLink and RouterLink components that are very similar in different parts of the application. I believe that we could merge all these components into one called Link or BaseLink.

I also noticed that, despite the different names, `ExternalLink` and` Link` have a very similar code..yes, we could have both, but then reusing the code of the another. 


-> Now most of our server-rendered views are gone, but there are still some remaining and this can break navigation on those views. We should try evaluate if we still need this behavior before removing it.

yes I am aware of it and this is the reason why I left the ExternalLink component.  I just checked and two views of getsentry are using links/link, so maybe I should leave the old link component as well...

**Comment by billyvg** [comment]
> yes I am aware of it and this is the reason why I left the ExternalLink component. I just checked and two views of getsentry are using links/link, so maybe I should leave the old link component as well...

It is not only the usage of `<Link>` and `<ExternalLink>` - it could be components that use these components. For example, I believe these are the "at-risk" components:

image

They are potentially affected because they are rendered without being wrapped by `<Router>` and thus do not have `this.context.location`. That means if the above components use `<Link>`, the links will break because there is no router. `<Link>` is used as a fallback for these scenarios (if they still exist) and the link is not treated as an "external" link, but a link to somewhere in our SPA but using normal browser navigation.


I would agree that `Link`, `RouterLink`, and `TextLink` are all very similar, but I think `ExternalLink` is a bit different, and I think it having a different API is beneficial. I also think typescript benefits the separate components approach more.

`<Link to="http://google.com" />` - if you forget the `external` prop, this becomes a broken link, typescript does not help here either

`<ExternalLink to="http://google.com" />` - typescript will tell you that `to` is an invalid property

**Comment by priscilawebdev** [comment]
@billyvg unfortunately the `React.ComponentProps<Link>` doesn't work because of the router thing... I will post the error here tomorrow

**Review by billyvg — COMMENTED** [review]
There seems to be a bunch of unrelated changes in here in an already complex refactor which makes it a bit more difficult to review :/

File: src/sentry/static/sentry/app/views/settings/settingsIndex.tsx [hunk 2]
```diff
@@ -324,11 +325,8 @@ const HomeLinkIcon = styled(HomeLink)`
 `;
 
 const ExternalHomeLink = styled(
-  ({
-    isCentered,
-    ...props
-  }: CenterableProps & React.ComponentProps<typeof ExternalLink>) => (
-    <ExternalLink {...props} />
+  (props: CenterableProps & React.ComponentProps<typeof ExternalLink>) => (
+    <ExternalLink {...omit(props, 'isCentered')} />
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/settings/settingsIndex.tsx]: Did this needed to be changed? On one hand it's a bit cleaner, on the other it pulls in lodash/omit.

File: src/sentry/static/sentry/app/views/settings/settingsIndex.tsx [hunk 3]
```diff
@@ -340,23 +338,20 @@ const ExternalHomeLink = styled(
...
+type SupportLinkProps<T extends boolean> = {
+  isOnPremise: T;
   href: string;
   to: string;
   isCentered?: boolean;
-} & (
-  | ({isOnPremise: true} & React.ComponentProps<typeof ExternalLink>)
-  | ({isOnPremise: false} & React.ComponentProps<typeof HomeLink>)
-);
+} & React.ComponentProps<T extends true ? typeof ExternalLink : typeof HomeLink>;
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/settings/settingsIndex.tsx]: This is much better!

File: src/sentry/static/sentry/app/views/settings/organizationDeveloperSettings/index.tsx [hunk 4]
```diff
@@ -139,7 +139,7 @@ class OrganizationDeveloperSettings extends AsyncView<Props, State> {
     return (
       <div>
         <SettingsPageHeader title={t('Developer Settings')} />
-        <AlertLink href="https://docs.sentry.io/workflow/integrations/integration-platform/">
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/settings/organizationDeveloperSettings/index.tsx]: I think we need to keep the `href` prop for the same reasoning discussed earlier - we need to be sure that `href` and `to` props are two very distinct things. People need to be aware of the differences between the two and having them combined like this can be confusing.

File: src/sentry/static/sentry/app/views/settings/components/teamSelect.jsx [hunk 5]
```diff
@@ -153,7 +153,7 @@ const TeamRow = props => {
   const {orgId, team, onRemove, disabled, confirmMessage} = props;
   return (
     <TeamPanelItem>
-      <StyledLink to={`/settings/${orgId}/teams/${team}/`}>#{team}</StyledLink>
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/settings/components/teamSelect.jsx]: Did this need to be changed?

File: src/sentry/static/sentry/app/views/organizationIntegrations/providerRow.tsx [hunk 6]
```diff
@@ -125,19 +124,19 @@ export default class ProviderRow extends React.Component<Props> {
   render() {
     return (
       <PanelItem p={0} flexDirection="column" data-test-id={this.props.provider.key}>
-        <Flex alignItems="center" p={2}>
+        <PanelItemLeft>
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationIntegrations/providerRow.tsx]: Thanks for cleaning this up

File: src/sentry/static/sentry/app/views/organizationIntegrations/providerRow.tsx [hunk 7]
```diff
@@ -125,19 +124,19 @@ export default class ProviderRow extends React.Component<Props> {
   render() {
     return (
       <PanelItem p={0} flexDirection="column" data-test-id={this.props.provider.key}>
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/organizationIntegrations/providerRow.tsx]: Can we also refactor this into a styled component? (Probably belongs in a new PR)

File: src/sentry/static/sentry/app/views/acceptOrganizationInvite.tsx [hunk 8]
```diff
@@ -76,7 +76,7 @@ class AcceptOrganizationInvite extends AsyncView<Props, State> {
           'Your account ([email]) is already a member of this organization. [switchLink:Switch accounts]?',
           {
             email: user.email,
-            switchLink: <Link href="#" onClick={this.handleLogout} />,
+            switchLink: <Link onClick={this.handleLogout} />,
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/acceptOrganizationInvite.tsx]: I think we should not have a default prop for `to`, it should be required and make sure people thinking about if it should be a link. A lot of times people use a Link when a Button is more appropriate.

That being said this should be `to="#"`

File: src/sentry/static/sentry/app/views/discover/result/table.tsx [hunk 9]
```diff
@@ -109,9 +109,7 @@ class ResultTable extends React.Component<Props> {
 
     return (
       <Tooltip title={t('Open event')}>
-        <Link href={`${basePath}events/${event.id}/`} target="_blank">
-          {event.id}
-        </Link>
+        <ExternalLink href={`${basePath}events/${event.id}/`}>{event.id}</ExternalLink>
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/discover/result/table.tsx]: @getsentry/visibility Are these two links intended to be opened in a new tab?

File: src/sentry/static/sentry/app/views/discover/aggregations/index.tsx [hunk 10]
```diff
@@ -58,9 +58,9 @@ export default class Aggregations extends React.Component<AggregationsProps> {
               disabled={disabled}
             />
             <div>
-              <a onClick={() => this.removeRow(idx)}>
+              <Link onClick={() => this.removeRow(idx)}>
```
Comment (billyvg) [inline src/sentry/static/sentry/app/views/discover/aggregations/index.tsx]: @getsentry/visibility Should this be a `Button`?

File: src/sentry/static/sentry/app/components/menuItem.tsx [hunk 11]
```diff
@@ -29,11 +29,6 @@ class MenuItem extends React.Component<Props> {
     eventKey: PropTypes.any,
     isActive: PropTypes.bool,
     noAnchor: PropTypes.bool,
-    // basic link
-    href: PropTypes.string,
-    // router link
-    to: PropTypes.string,
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/menuItem.tsx]: Can we move these changes to a separate PR? I think we should focus on just `Link`

File: src/sentry/static/sentry/app/components/alertMessage.tsx [hunk 12]
```diff
@@ -37,7 +37,7 @@ const AlertMessage = ({alert, system}: Props) => {
         onClick={handleCloseAlert}
         borderless
       />
-      {url ? <Link href={url}>{message}</Link> : message}
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/alertMessage.tsx]: These might need to be <ExternalLink>. I'm not sure if and how they are used.

**Review by billyvg — COMMENTED** [review]
Thanks so much for working on this, having this in typescript will hopefully clarify how to use Link vs ExternalLink more

File: src/sentry/static/sentry/app/components/alertLink.tsx [hunk 13]
```diff
@@ -50,16 +47,16 @@ export default class AlertLink extends React.Component<Props> {
       onClick,
       withoutMarginBottom,
       openInNewTab,
+      to,
     } = this.props;
     return (
       <StyledLink
-        to={(this.props as PropsWithTo).to}
-        href={(this.props as PropsWithHref).href}
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/alertLink.tsx]: As mentioned in https://github.com/getsentry/sentry/pull/17346#discussion_r395202569 - I think these need to remain.

I also think that it was intentional here to open an external link but still have it open in the same tab since it was linking to our docs.

ExternalLink might need a prop to disable opening in a new tab (default should open in a new tab).

File: src/sentry/static/sentry/app/components/links/linkWithConfirmation.tsx [hunk 14]
```diff
@@ -1,39 +1,35 @@
...
-    };
-  }
-
+class LinkWithConfirmation extends React.PureComponent<Props> {
   render() {
     const {className, disabled, title, children, ...otherProps} = this.props;
     return (
       <Confirm {...otherProps} disabled={disabled}>
-        <a
+        <Link
```
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/links/linkWithConfirmation.tsx]: I believe that if this is not a link, we should not use the `<a>` tag here, but style the container in a way that looks like a link.. but it's true...we are using the Link component in many places without `href` or `to`, but we should replace them with a Button component...but not sure if we should do all the changes in this PR

**Review by priscilawebdev — COMMENTED** [review]
@billyvg thank you for reviewing. what do you think about solving this together during a pair programming section? I believe it would more efficient and we could soon have this merged :)

File: src/sentry/static/sentry/app/components/sidebar/sidebarItem.tsx [hunk 15]
```diff
@@ -130,7 +128,7 @@ const SidebarItem = ({
 
 export default ReactRouter.withRouter(SidebarItem);
 
-const getActiveStyle = ({active, theme}) => {
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/sidebar/sidebarItem.tsx]: We should try to avoid using imported `theme` when possible, using `theme` from props is preferred because it's through emotion's `ThemeProvider` - this is how we can have multiple themes (if we ever add multiple themes)

File: src/sentry/static/sentry/app/components/sidebar/sidebarMenuItem.tsx [hunk 16]
```diff
@@ -4,23 +4,22 @@ import {css} from '@emotion/core';
...
 type Props = {
   children: React.ReactNode;
-} & SidebarMenuItemLinkProps;
+} & React.ComponentProps<typeof SidebarMenuItemLink>;
 
-const SidebarMenuItem = ({to, href, children, ...props}: Props) => {
-  const hasMenu = !to && !href;
-  return (
-    <StyledSidebarMenuItemLink to={to} href={href} {...props}>
-      <MenuItemLabel hasMenu={hasMenu}>{children}</MenuItemLabel>
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/sidebar/sidebarMenuItem.tsx]: We need to restore this because of https://github.com/getsentry/sentry/pull/17346/files#diff-4734084577830b057fc992bcdc606d26R44

File: tests/acceptance/test_api.py [hunk 17]
```diff
@@ -42,8 +42,6 @@ def test_simple(self):
 
         self.browser.click('[href="/settings/account/api/applications/"]')
         self.browser.wait_until_not(".loading")
-        self.browser.click_when_visible('[data-test-id="toast-success"]')
-        self.browser.wait_until_not('[data-test-id="toast-success"]')
```
Comment (billyvg) [inline tests/acceptance/test_api.py]: Why remove this?

File: tests/js/spec/components/sidebar/__snapshots__/index.spec.jsx.snap [hunk 18]
```diff
@@ -989,75 +974,55 @@ exports[`Sidebar SidebarHelp can toggle help menu 1`] = `
...
-                >
-                  Community Discussions
-                </span>
-              </MenuItemLabel>
-            </a>
-          </Link>
+                Community Discussions
+              </span>
+            </MenuItemLabel>
+          </div>
```
Comment (billyvg) [inline tests/js/spec/components/sidebar/__snapshots__/index.spec.jsx.snap]: Hmm is this right? I would expect to see `<ExternalLink>` somewhere.

File: src/sentry/static/sentry/app/components/links/externalLink.tsx [hunk 19]
```diff
@@ -1,8 +1,32 @@
...
-type AnchorProps = React.HTMLProps<HTMLAnchorElement>;
-type Props = AnchorProps & Required<Pick<AnchorProps, 'href'>>;
+type Props = {
+  // Link URL
+  href: string;
+  // Link content (accepted via string or components / DOM nodes)
+  children?: React.ReactNode;
+  // Styles applied to the component's root
+  className?: string;
+  // Action to perform when clicked (will cause the component to be rendered as a button instead of an anchor)
```
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/links/externalLink.tsx]: Since now most of the props comes from the AnchorProps, I removed the comments

File: tests/js/spec/components/sidebar/__snapshots__/index.spec.jsx.snap (hunk 18 above)
Comment (priscilawebdev) [reply tests/js/spec/components/sidebar/__snapshots__/index.spec.jsx.snap]: good catch..return was missing ;-)

File: src/sentry/static/sentry/app/components/menuItem.tsx [hunk 20]
```diff
@@ -58,18 +52,6 @@ class MenuItem extends React.Component<Props> {
...
-        <a
-          title={this.props.title}
-          onClick={this.handleClick}
-          href={this.props.href}
-          tabIndex={-1}
-        >
-          {this.props.children}
-        </a>
-      );
-    }
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/menuItem.tsx]: Are we sure this isn't in use? Can we be a bit more defensive restore this code and do something similar in `Link` where we track to Sentry if `this.props.href` exists. My fear is that this is still in use somewhere and we do not have any visibility if it does since it will not cause any errors but it will still appear broken to the user.

File: src/sentry/static/sentry/app/components/modals/sentryAppDetailsModal.tsx [hunk 21]
```diff
@@ -213,7 +208,22 @@ export default class SentryAppDetailsModal extends AsyncComponent<Props, State>
   }
 }
 
-const Name = styled(Box)`
+const DetailsWrapper = styled('div')`
+  display: flex;
+  margin-bottom: 2px;
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/modals/sentryAppDetailsModal.tsx]: These pixel values are not correct, it should be `space(2)`, but this is another reason why these changes should not be in this PR.

**Review by billyvg — COMMENTED** [review]
Can we pull out the unrelated changes from this PR? This is already a big and risky PR and having these unrelated changes makes it harder for reviewers to review, and just adds more things that can go wrong.

File: src/sentry/static/sentry/app/components/alertLink.tsx [hunk 22]
```diff
@@ -69,9 +67,23 @@ export default class AlertLink extends React.Component<Props> {
...
+export default AlertLink;
+
+const StyledLink = styled(({openInNewTab, to, href, ref, ...props}: StyledLinkProps) => {
+  const linkProps = omit(props, ['withoutMarginBottom', 'priority', 'size']);
+  if (href) {
+    return (
+      <ExternalLink
+        {...linkProps}
+        href={href}
+        ref={ref as React.RefObject<HTMLAnchorElement>}
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/alertLink.tsx]: I don't think this is needed (or works as you expect)

Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/alertLink.tsx]: As we are now using `type AnchorProps = React.HTMLProps<HTMLAnchorElement>` I had some issues with the ref and I decided to do it like this for now ...

File: src/sentry/static/sentry/app/components/modals/sentryAppDetailsModal.tsx (hunk 21 above)
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/modals/sentryAppDetailsModal.tsx]: I only created these components, because lint was not passing because of the reflexbox import and I didn't want to do many changes, because it doesn't belong to this PR.

File: src/sentry/static/sentry/app/views/eventsV2/sortLink.tsx [hunk 23]
```diff
@@ -59,11 +59,12 @@ class SortLink extends React.Component<Props> {
...
 }
 
-type StyledLinkProps = Link['props'] & {align: Alignments};
+type LinkProps = React.ComponentProps<typeof Link>;
+type StyledLinkProps = LinkProps & {align: Alignments};
 
 const StyledLink = styled((props: StyledLinkProps) => {
-  const forwardProps = omit(props, ['align']);
-  return <Link {...forwardProps} />;
+  const {ref, ...forwardProps} = omit(props, ['align']);
```
[the remark this answers is not in the document]
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/eventsV2/sortLink.tsx]: I was having some issues with the ref and decided to do like `ref={ref as any} ` for now

[the remark this answers is not in the document]
Comment (priscilawebdev) [reply src/sentry/static/sentry/app/views/eventsV2/sortLink.tsx]: thanks for the `React.ComponentWithoutRef` tip :)

**Comment by billyvg** [comment]
Closes #16429

File: src/sentry/static/sentry/app/components/links/externalLink.tsx [hunk 24]
```diff
@@ -1,8 +1,31 @@
...
+const ExternalLink = React.forwardRef<HTMLAnchorElement, Props>(function ExternalLink(
+  {openInNewTab = true, ...props},
+  ref
+) {
+  const anchorProps = openInNewTab ? {target: '_blank', rel: 'noreferrer noopener'} : {};
+  return (
+    <a
+      ref={ref}
+      target={openInNewTab ? '_blank' : '_self'}
+      rel="noreferrer noopener"
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/links/externalLink.tsx]: We can remove these now

```suggestion
```

File: src/sentry/static/sentry/app/components/sidebar/sidebarMenuItemLink.tsx [hunk 25]
```diff
@@ -3,23 +3,17 @@ import React from 'react';
...
    */
   href?: string;
-  /**
-   * Is an external link? (Will open in new tab; Only applicable if `href` is used)
-   */
-  external?: boolean;
-  /**
-   * specifies whether to open the linked document in a new tab
-   */
-  openInNewTab?: boolean;
```
Comment (billyvg) [inline src/sentry/static/sentry/app/components/sidebar/sidebarMenuItemLink.tsx]: This can get passed through to `<ExternalLink>` now, correct?

Comment (priscilawebdev) [reply src/sentry/static/sentry/app/components/sidebar/sidebarMenuItemLink.tsx]: yes

**Review by billyvg — APPROVED** [review]
Clicked around on staging and things seemed to work well. 🎉
